File: frontend-react/src/api/apiClient.ts
import {
  buildCreateBody,
  buildPromoteBody,
  buildPromoteBodyImplicit,
  buildRejectBody,
  buildResolveBody,
  buildUpdateBody,
  contractHeaders,
} from "./contract";
import {
  candidateFromMapping,
  parseCandidateList,
} from "./candidateModel";
import {
  deriveGraphFromCandidates,
  parseGraphPayload,
} from "./graphModel";
import type { DataProvider } from "./dataProvider";
import type { CandidateWriteInput, EvalMetrics } from "../types/candidate";

class ApiConflictError extends Error {
  readonly statusCode = 409 as const;
  candidateId?: string;

  constructor(message: string, candidateId?: string) {
    super(message);
    this.name = "ApiConflictError";
    this.candidateId = candidateId;
  }
}

class ApiClientError extends Error {
  statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = "ApiClientError";
    this.statusCode = statusCode;
  }
}

function extractCandidateId(path: string): string | undefined {
  const prefix = "/candidates/";
  if (!path.includes(prefix)) {
    return undefined;
  }
  const segment = path.split(prefix)[1]?.split("/")[0];
  return segment ? decodeURIComponent(segment) : undefined;
}

function isPromoteConflict(error: ApiClientError): boolean {
  return (
    error.statusCode === 400 &&
    error.message.toLowerCase().includes("cannot promote")
  );
}

function toPromoteConflict(error: ApiClientError, candidateId: string): ApiConflictError {
  return new ApiConflictError(error.message, candidateId);
}

async function parseJsonResponse(response: Response): Promise<unknown> {
  const raw = await response.text();
  if (!raw.trim()) {
    return {};
  }
  return JSON.parse(raw) as unknown;
}

export interface ApiDataProviderAuth {
  /** Resolve a currently-valid bearer token (Amplify refreshes on demand). */
  getToken?: () => Promise<string | undefined>;
  /** Active org id sent as X-Praxis-Org for server-side tenancy. */
  orgId?: string;
}

export function createApiDataProvider(
  baseUrl: string,
  auth?: ApiDataProviderAuth,
  evalMetricsUrl?: string,
): DataProvider {
  const root = baseUrl.replace(/\/$/, "");
  const metricsUrl =
    evalMetricsUrl?.trim() ||
    import.meta.env.VITE_PRAXIS_EVAL_METRICS_URL?.trim() ||
    `${root}/metrics`;

  async function authHeaders(): Promise<HeadersInit> {
    const token = auth?.getToken ? await auth.getToken() : undefined;
    return contractHeaders(token, auth?.orgId);
  }

  async function request(
    method: string,
    path: string,
    body?: Record<string, unknown>,
  ): Promise<unknown> {
    const response = await fetch(`${root}${path}`, {
      method,
      headers: await authHeaders(),
      body: body ? JSON.stringify(body) : undefined,
    });

    if (!response.ok) {
      const detail = await response.text();
      if (response.status === 409) {
        throw new ApiConflictError(
          detail || response.statusText,
          extractCandidateId(path),
        );
      }
      throw new ApiClientError(
        `API ${method} ${path} failed (${response.status}): ${detail || response.statusText}`,
        response.status,
      );
    }

    return parseJsonResponse(response);
  }

  return {
    async listCandidates(state?: string) {
      const query = state ? `?state=${encodeURIComponent(state)}` : "";
      const payload = await request("GET", `/candidates${query}`);
      return parseCandidateList(payload).map(candidateFromMapping);
    },

    async getCandidate(id) {
      try {
        const payload = await request(
          "GET",
          `/candidates/${encodeURIComponent(id)}`,
        );
        if (payload && typeof payload === "object") {
          return candidateFromMapping(payload as Record<string, unknown>);
        }
        return null;
      } catch (error) {
        if (error instanceof ApiClientError && error.statusCode === 404) {
          return null;
        }
        throw error;
      }
    },

    async promote(id) {
      const current = await this.getCandidate(id);
      if (!current) {
        throw new Error(`Unknown candidate id: ${id}`);
      }

      const path = `/candidates/${encodeURIComponent(id)}/promote`;
      try {
        const payload = await request("POST", path, buildPromoteBody(current.state));
        return candidateFromMapping(payload as Record<string, unknown>);
      } catch (error) {
        if (error instanceof ApiClientError && isPromoteConflict(error)) {
          throw toPromoteConflict(error, id);
        }
        if (
          error instanceof ApiClientError &&
          (error.statusCode === 400 || error.statusCode === 422)
        ) {
          try {
            const payload = await request("POST", path, buildPromoteBodyImplicit());
            return candidateFromMapping(payload as Record<string, unknown>);
          } catch (retryError) {
            if (
              retryError instanceof ApiClientError &&
              isPromoteConflict(retryError)
            ) {
              throw toPromoteConflict(retryError, id);
            }
            throw retryError;
          }
        }
        throw error;
      }
    },

    async reject(id, reason) {
      await request(
        "POST",
        `/candidates/${encodeURIComponent(id)}/reject`,
        buildRejectBody(reason),
      );
    },

    async createCandidate(input: CandidateWriteInput) {
      const payload = await request("POST", "/candidates", buildCreateBody(input));
      return candidateFromMapping(payload as Record<string, unknown>);
    },

    async updateCandidate(id, input) {
      const payload = await request(
        "PATCH",
        `/candidates/${encodeURIComponent(id)}`,
        buildUpdateBody(input),
      );
      return candidateFromMapping(payload as Record<string, unknown>);
    },

    async deleteCandidate(id) {
      await request("DELETE", `/candidates/${encodeURIComponent(id)}`);
    },

    async resolveContradiction(contradictionId, resolution, keepId) {
      const payload = await request(
        "POST",
        `/contradictions/${encodeURIComponent(contradictionId)}/resolve`,
        buildResolveBody(resolution, keepId),
      );
      return candidateFromMapping(payload as Record<string, unknown>);
    },

    async getEvalMetrics() {
      try {
        const response = await fetch(metricsUrl, {
          headers: await authHeaders(),
        });
        if (!response.ok) {
          throw new Error(response.statusText);
        }
        const payload = (await response.json()) as Record<string, unknown>;
        return normalizeEvalMetrics(payload, metricsUrl);
      } catch (error) {
        const message =
          error instanceof Error ? error.message : "Eval metrics unavailable";
        return {
          source: "placeholder",
          correctionRate: [1.0, 0.72, 0.48, 0.35],
          sessions: ["cold", "run_1", "run_2", "run_3"],
          correctionsBefore: 12,
          correctionsAfter: 5,
          fetchError: message,
        };
      }
    },

    async getGraph() {
      try {
        const payload = await request("GET", "/graph");
        return parseGraphPayload(payload, "api");
      } catch (error) {
        if (
          error instanceof ApiClientError &&
          (error.statusCode === 404 || error.statusCode === 405)
        ) {
          const rows = await this.listCandidates();
          return deriveGraphFromCandidates(rows);
        }
        if (error instanceof ApiClientError) {
          const rows = await this.listCandidates();
          return deriveGraphFromCandidates(rows);
        }
        throw error;
      }
    },

    async getTranscript() {
      return null;
    },
  };
}

export async function postIngestJsonl(
  apiBaseUrl: string,
  files: Array<{ name: string; content: string }>,
  auth?: string | ApiDataProviderAuth,
): Promise<void> {
  const root = apiBaseUrl.replace(/\/$/, "");
  const resolved: ApiDataProviderAuth =
    typeof auth === "string" ? { getToken: async () => auth } : auth ?? {};
  const token = resolved.getToken ? await resolved.getToken() : undefined;
  const response = await fetch(`${root}/ingest/jsonl`, {
    method: "POST",
    headers: contractHeaders(token, resolved.orgId),
    body: JSON.stringify({ files }),
  });

  if (!response.ok) {
    const detail = await response.text();
    if (response.status === 404 || response.status === 405) {
      throw new Error("Distillation endpoint not available yet");
    }
    throw new ApiClientError(
      `API POST /ingest/jsonl failed (${response.status}): ${detail || response.statusText}`,
      response.status,
    );
  }
}

function normalizeEvalMetrics(
  payload: Record<string, unknown>,
  source: string,
): EvalMetrics {
  const series =
    (payload.correction_rate as number[] | undefined) ??
    (payload.correctionRate as number[] | undefined) ??
    [];
  const sessions = payload.sessions as string[] | undefined;
  const correctionsBefore =
    (payload.corrections_before as number | undefined) ??
    (payload.correctionsBefore as number | undefined);
  const correctionsAfter =
    (payload.corrections_after as number | undefined) ??
    (payload.correctionsAfter as number | undefined);

  return {
    source,
    correctionRate: Array.isArray(series) ? series : [],
    sessions,
    correctionsBefore,
    correctionsAfter,
  };
}

export { ApiClientError, ApiConflictError };

File: frontend/services/api_client.py
"""
===============================================================================
FILE: services/api_client.py
AUTHOR: Monica Peters
CREATED: 2026-06-18

PURPOSE:
HTTP DataProvider for Matthew's pipeline API (Days 6–7 integration).

USAGE:
    provider = ApiDataProvider(base_url=os.environ["PRAXIS_API_BASE_URL"])

SECURITY:
- Token via PRAXIS_API_TOKEN environment variable only.

OPERATIONAL:
- Targets docs/integration/candidate-api-v1.md (contract v1).
- Uses stdlib urllib — no extra HTTP dependencies.
- No imports from pipeline/ or eval/ — Matthew owns server implementation.
===============================================================================
"""

from __future__ import annotations

import json
import logging
import urllib.error
import urllib.parse
import urllib.request
from typing import Any

from models.candidate import Candidate, CandidateState
from services.contract_v1 import (
    build_promote_body,
    build_promote_body_implicit,
    build_reject_body,
    build_resolve_body,
    contract_headers,
    parse_candidate_list,
)

logger = logging.getLogger(__name__)


class ApiConflictError(Exception):
    """Raised when the API returns HTTP 409 on a mutation."""

    def __init__(self, message: str, *, candidate_id: str | None = None) -> None:
        super().__init__(message)
        self.candidate_id = candidate_id


class ApiClientError(Exception):
    """Raised when the API returns a non-success HTTP status."""

    def __init__(self, message: str, *, status_code: int) -> None:
        super().__init__(message)
        self.status_code = status_code


class ApiDataProvider:
    """
    Thin HTTP client over Matthew's REST API.

    Python reference client for candidate-api-v1 — returns typed Candidate models.
    The React dashboard in frontend-react/ calls the same endpoints; the contract
    itself is UI-agnostic.
    """

    def __init__(
        self,
        base_url: str,
        token: str | None = None,
        org_id: str | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._token = token
        self._org_id = org_id

    def list_candidates(self, state: CandidateState | None = None) -> list[Candidate]:
        query = ""
        if state is not None:
            query = "?" + urllib.parse.urlencode({"state": state.value})
        payload = self._request("GET", f"/candidates{query}")
        rows = parse_candidate_list(payload)
        return [Candidate.from_mapping(row) for row in rows]

    def get_candidate(self, candidate_id: str) -> Candidate | None:
        try:
            payload = self._request("GET", f"/candidates/{urllib.parse.quote(candidate_id, safe='')}")
        except ApiConflictError:
            raise
        except ApiClientError as exc:
            if exc.status_code == 404:
                return None
            raise
        if isinstance(payload, dict):
            return Candidate.from_mapping(payload)
        return None

    def promote(self, candidate_id: str) -> Candidate:
        encoded = urllib.parse.quote(candidate_id, safe="")
        path = f"/candidates/{encoded}/promote"
        current = self.get_candidate(candidate_id)
        if current is None:
            raise KeyError(f"Unknown candidate id: {candidate_id!r}")

        explicit_body = build_promote_body(current_state=current.state)
        try:
            payload = self._request("POST", path, body=explicit_body)
        except ApiClientError as exc:
            if _is_promote_conflict(exc):
                raise ApiConflictError(
                    f"Conflict: {exc}",
                    candidate_id=candidate_id,
                ) from exc
            if exc.status_code not in (400, 422):
                raise
            logger.info(
                "Promote with targetState rejected (%s); retrying with implicit body",
                exc.status_code,
            )
            try:
                payload = self._request("POST", path, body=build_promote_body_implicit())
            except ApiClientError as retry_exc:
                if _is_promote_conflict(retry_exc):
                    raise ApiConflictError(
                        f"Conflict: {retry_exc}",
                        candidate_id=candidate_id,
                    ) from retry_exc
                raise

        if not isinstance(payload, dict):
            raise ValueError("Promote response must be a candidate object")
        return Candidate.from_mapping(payload)

    def reject(self, candidate_id: str, reason: str | None = None) -> None:
        encoded = urllib.parse.quote(candidate_id, safe="")
        self._request(
            "POST",
            f"/candidates/{encoded}/reject",
            body=build_reject_body(reason=reason),
        )

    def resolve_contradiction(
        self,
        contradiction_id: str,
        *,
        resolution: str,
        keep_id: str,
    ) -> Candidate:
        encoded = urllib.parse.quote(contradiction_id, safe="")
        body = build_resolve_body(resolution=resolution, keep_id=keep_id)
        payload = self._request("POST", f"/contradictions/{encoded}/resolve", body=body)
        if not isinstance(payload, dict):
            raise ValueError("Resolve response must include the kept candidate")
        return Candidate.from_mapping(payload)

    def _headers(self) -> dict[str, str]:
        return contract_headers(token=self._token, org_id=self._org_id)

    def _request(
        self,
        method: str,
        path: str,
        *,
        body: dict[str, Any] | None = None,
    ) -> Any:
        url = f"{self._base_url}{path}"
        data = None
        if body is not None:
            data = json.dumps(body).encode("utf-8")
        request = urllib.request.Request(url, data=data, headers=self._headers(), method=method)
        try:
            with urllib.request.urlopen(request, timeout=30) as response:
                raw = response.read().decode("utf-8")
                if not raw.strip():
                    return {}
                return json.loads(raw)
        except urllib.error.HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")
            if exc.code == 409:
                raise ApiConflictError(
                    f"Conflict (409): {detail or exc.reason}",
                    candidate_id=_extract_candidate_id(path),
                ) from exc
            if exc.code in (401, 403):
                # The server requires a Cognito Bearer JWT (PRAXIS_API_TOKEN) and an
                # org the caller belongs to (PRAXIS_ORG_ID -> X-Praxis-Org). Surface
                # that explicitly instead of a generic API error.
                hint = (
                    "missing or invalid bearer token"
                    if exc.code == 401
                    else "token valid but not a member of the requested org"
                )
                raise ApiClientError(
                    f"API {method} {path} unauthorized ({exc.code}): {hint}. "
                    f"Set PRAXIS_API_TOKEN and PRAXIS_ORG_ID (or PRAXIS_AUTH_DISABLED=1 "
                    f"on a dev server). Detail: {detail or exc.reason}",
                    status_code=exc.code,
                ) from exc
            raise ApiClientError(
                f"API {method} {path} failed ({exc.code}): {detail or exc.reason}",
                status_code=exc.code,
            ) from exc
        except urllib.error.URLError as exc:
            raise RuntimeError(f"API unreachable: {exc.reason}") from exc


def _extract_candidate_id(path: str) -> str | None:
    prefix = "/candidates/"
    if prefix not in path:
        return None
    remainder = path.split(prefix, 1)[1]
    segment = remainder.split("/", 1)[0]
    return urllib.parse.unquote(segment) if segment else None


def _is_promote_conflict(exc: ApiClientError) -> bool:
    """Matthew's server returns 400 PromotionError for stale/invalid promote."""
    if exc.status_code != 400:
        return False
    return "cannot promote" in str(exc).lower()
